src/main/java/org/moonlightcontroller/aggregator/ApplicationAggregator.java

```diff
@@ -1,146 +1,718 @@
+				new Connector.Builder().setSourceBlock(lbBlocks[1]).setSourceOutputPort(1).setDestBlock(lbBlocks[4]).build(),
+				new Connector.Builder().setSourceBlock(lbBlocks[2]).setSourceOutputPort(0).setDestBlock(lbBlocks[3]).build(),
+				new Connector.Builder().setSourceBlock(lbBlocks[3]).setSourceOutputPort(0).setDestBlock(lbBlocks[4]).build(),
+		};
+		
+		IProcessingGraph a = new ProcessingGraph.Builder()
+			.setBlocks(ImmutableList.copyOf(firewallBlocks))
+			.setConnectors(ImmutableList.copyOf(firewallConnectors))
+			.setRoot(firewallBlocks[0])
+			.build();
+		
+		System.out.println("Input graph A:");
+		System.out.println(a);
+		
+		IProcessingGraph b = new ProcessingGraph.Builder()
+			.setBlocks(ImmutableList.copyOf(lbBlocks))
+			.setConnectors(ImmutableList.copyOf(lbConnectors))
+			.setRoot(lbBlocks[0])
+			.build();
+		
+		System.out.println("Input graph B:");
+		System.out.println(b);
+		
+		IProcessingGraph merged = merge(a, b);
+		
+		System.out.println("Result graph:");
+		System.out.println(merged);
+		
+		System.out.println();
+		
+		for (IProcessingBlock block : merged.getBlocks()) {
+			if (block instanceof HeaderClassifier) {
+				HeaderClassifier classifier = (HeaderClassifier)block;
+				printRules(classifier);
+			}
+		}
+		
+		System.out.println();
+		System.out.println("Total number of blocks: " + merged.getBlocks().size());
+	}
+	
+	private static void printRules(HeaderClassifier classifier) {
+		int i = 0;
+		System.out.println("Rules for classifier " + classifier.getId() + ":");
+		for (HeaderClassifierRule rule: classifier.getRules()) {
+			System.out.println(String.format("%s --> [%d]", rule.getMatch(), i++));
+		}
+	}
+
+	private static ApplicationAggregator _instance;
+	
+	private List<BoxApplication> apps;
+	private Map<ILocationSpecifier, IProcessingGraph> aggregated;
+	
+	private ApplicationAggregator() {
+		this.apps = new ArrayList<>();
+		this.aggregated = new HashMap<>();
+	}
+	
+	public static IApplicationAggregator getInstance() {
+		if (_instance == null) {
+			synchronized (ApplicationAggregator.class) {
+				if (_instance == null) {
+					_instance = new ApplicationAggregator();
+				}
+			}
+		}
+		return _instance;
+	}
+	
+	@Override
+	public void addApplications(List<BoxApplication> apps) {
+		this.apps.addAll(apps);
+	}
+
+	@Override
+	public void addApplication(BoxApplication app) {
+		this.apps.add(app);
+	}
+
+	@Override
+	public void performAggregation() {
+		synchronized (this) {
+			// TODO: The following should be done for each OBI location specifier
+			List<InstanceLocationSpecifier> obiLocationSpecifiers = TopologyManager.getInstance().getAllEndpoints(); 
+			for (InstanceLocationSpecifier loc : obiLocationSpecifiers) {
+				IProcessingGraph last = null;
+				for (BoxApplication app : this.apps) {
+					// TODO: Complete here
+					// current = app.getProcessingGraph(); // This method does not exist now
```

Comment:
Not sure what should be done here.
Why should it be done for each OBI.
I think it's better to not block this merge on it.
I will do it in my next merge.